docs/upgrading/v6-data.md

```diff
@@ -0,0 +1,216 @@
+---
+title: Migrating to RouterProvider
+order: 1
+---
+
+# Migrating to RouterProvider
+
+When we originally began bringing the [Remix Data APIs over to React Router][remixing-react-router] we realized that they brought about a pretty different way of structuring your routes. Instead of discovering routes via the [`<Routes>`][routes-component] component _as React rendered the component tree_, we needed to _lift_ the route definitions so we could [decouple fetching from rendering][when-to-fetch].
+
+This brought about an interesting conundrum. We've got tons of v6 [`BrowserRouter`][browserrouter] apps out there happily defining their routes via `<Routes>` components -- how can we provide them with a smooth upgrade experience that doesn't require a big-bang migration to the new approach? This ruled out a new major version and we focused on adding these new features in a _fully backwards compatible_ way that would provide users an _incremental_ upgrade path from [`BrowserRouter`][browserrouter] to [`RouterProvider`][routerprovider].
+
+## Differences
+
+The first thing to be aware of is the presence of a handful of new [Data APIs][data-apis] that only work on routes _defined_ via the new [data routers][picking-a-router] (i.e., `createBrowserRouter`). These include a few categories of APIs:
+
+- Route-level data APIs such as `loader`, `action`, `shouldRevalidate`, `handle`, and `lazy`
+- In-component data hooks such as `useLoaderData`, `useActionData`, `useFetcher`, `useMatches`, `useNavigation`, etc.
+- Error-handling APIs such as `route.errorElement`, `route.ErrorBoundary`, and `useRouteError`
+
+The rest of the APIs that existed prior to v6.4.0 are still usable in _both_ `BrowserRouter` and `RouterProvider` apps. These include common hooks/components such as `useNavigate`, `useLocation`, `useParams`, `<Link>`, `<Outlet />`, etc.
+
+## Migrating
+
+We built the new `<RouterProvider>` component such that it would enable the new Data APIs on routes defined at the root router, while not excluding descendant `<Routes>` trees so commonly used in `BrowserRouter` apps. This was explicitly to allow incremental migration from one to the other. Let's take a look at how we would do this.
+
+### Current Application
+
+Let's assume we've got a current application with 2 descendant route trees, and assume these routes are all doing in-component data fetching, and rendering their own loading and error states.
+
+```tsx
+import * as React from "react";
+import {
+  BrowserRouter,
+  Link,
+  Route,
+  Routes,
+} from "react-router-dom";
+
+export default function App() {
+  return (
```

Comment:
it would be great to provide an example where App is doing more setup code w/ hooks - you'll need to either pull that up outside of the router or into a layout component at the router root